File: 560-subarray-sum-equals-k/subarray-sum-equals-k.cpp
class Solution {
public:
    int subarraySum(vector<int>& nums, int k) {
        int count = 0 ;
        int n = nums.size();


        // brtue tha 
        // for(int i = 0 ; i < n ; i++){
        //     int sum = 0 ;
        //     for(int j = i ; j < n ; j++){
        //         sum += nums[j] ;
        //         if(sum == k)  count++;
        //     }
        // }


        unordered_map<int,int> mpp;
        int sum = 0 ;
        mpp[0] = 1 ;

        for(int i = 0 ; i < n ; i++){

            sum += nums[i];

            int result = sum - k ;
            if(mpp.find(result) != mpp.end()){
                count+=mpp[result];
            }

            mpp[sum]++;
        }


        
        
        return count;
    }
};

File: 560-subarray-sum-equals-k/subarray-sum-equals-k.java
class Solution {
    public int subarraySum(int[] nums, int k) {

        HashMap<Integer, Integer> mpp = new HashMap<>() ;
        int sum = 0 ; 
        int total = 0 ; 

        mpp.put(0 , 1);
        
        for(int i : nums){

            sum += i ; 

            if(mpp.containsKey(sum - k )){
                total += mpp.get(sum - k);
            }

            mpp.put(sum , mpp.getOrDefault(sum , 0)+1);

        } 

        return total ; 

    }
}